### PR TITLE
Make progress bar "backend" configurable, use the "log" one during tests

### LINKED ISSUE #3393: Alternatives to progress bars
#### What is the problem?

The current progress bar works well in a proper terminal, but if running in a less well-endowed environment, you can end up with something like:

```
    Total:   0%|                                         | 0.00/647M [00:00<?, ?B/s]
    sub-01/func .. bold.nii.gz:   0%|                    | 0.00/212M [00:00<?, ?B/s][A
    
    sub-01/func .. ressors.tsv:   0%|                   | 0.00/65.1k [00:00<?, ?B/s][A[A
    
    
    sub-02/func .. bold.nii.gz:   0%|                    | 0.00/212M [00:00<?, ?B/s][A[A[A
    
    
    
    Total:   0%|                              | 65.1k/647M [00:01<2:58:41, 60.3kB/s]
    
    Total (1 ok out of 6):   0%|              | 65.1k/647M [00:01<2:58:41, 60.3kB/s]
    
    Total (1 ok out of 6):   1%|▏             | 7.42M/647M [00:01<2:03:45, 86.1kB/s]
    sub-01/func .. bold.nii.gz:   3%|▍          | 7.35M/212M [00:00<00:14, 14.6MB/s][A
    
    sub-02/func .. bold.nii.gz:   5%|▌          | 9.70M/212M [00:00<00:10, 19.3MB/s][A[A
    
    
    Total (3 ok out of 6):   4%|▌              | 26.4M/647M [00:01<1:24:06, 123kB/s]
    sub-01/func .. bold.nii.gz:   4%|▍          | 7.80M/212M [00:01<02:26, 1.39MB/s][A
    
    sub-02/func .. bold.nii.gz:   5%|▌          | 10.5M/212M [00:01<01:24, 2.38MB/s][A[A
    
    
    Total (3 ok out of 6):   5%|▋              | 29.1M/647M [00:03<1:01:06, 168kB/s]
    sub-01/func .. bold.nii.gz:   4%|▍           | 8.31M/212M [00:03<04:43, 718kB/s][A
    
    sub-02/func .. bold.nii.gz:   5%|▌           | 11.0M/212M [00:03<03:43, 901kB/s][A[A
    
    
    sub-03/func .. bold.nii.gz:   5%|▌           | 10.2M/223M [00:03<04:31, 783kB/s][A[A[A
    sub-01/func .. bold.nii.gz:   4%|▍           | 8.57M/212M [00:03<05:16, 642kB/s][A
    
    
    sub-03/func .. bold.nii.gz:   5%|▌           | 10.5M/223M [00:03<04:50, 730kB/s][A[A[A
    sub-01/func .. bold.nii.gz:   4%|▍           | 8.77M/212M [00:04<06:09, 549kB/s][A
    
    Total (3 ok out of 6):   5%|▊                | 31.2M/647M [00:04<44:09, 232kB/s]
```

I've looked, and there doesn't seem to be any way to select an alternative display, or even no progress display.

It would be nice to be able to set a configuration option `datalad.progressbar = log` or similar to indicate that output should be treated as a serial stream, and `datalad.progressbar = none` to simply not try to display the information.

#### What steps will reproduce the problem?

In `jupyter-lab` with a Bash kernel:

```
datalad install -r ///labs/poldrack/ds003_fmriprep
datalad get ds003_fmriprep/sub-0{1,2,3}/func/*_space-MNI152NLin2009cAsym_desc-*.nii.gz \
            ds003_fmriprep/sub-0{1,2,3}/func/*_desc-confounds_*.tsv
```

#### What version of DataLad are you using (run `datalad --version`)? On what operating system (consider running `datalad wtf`)?

datalad 0.11.4

#### Have you had any success using DataLad before? (to assess your expertise/prior luck.  We would welcome your testimonial additions to https://github.com/datalad/datalad/wiki/Testimonials as well)

Yes

### PR DESCRIPTION
**Pull request by yarikoptic**
Closes #3393

Adds `LogProgressBar` which logs completion (possibly partial) for the item.

<details>
<summary>Here are examples on how reporting looks like with `log` with some being interrupted or completed fully</summary>

```shell
$> sudo rm -rf QA simon ds00* face_place dartmouth-phantoms; DATALAD_UI_PROGRESSBAR=log datalad install -g ///openfmri/ds000105
[INFO   ] Cloning http://datasets.datalad.org/openfmri/ds000105 [1 other candidates] into '/tmp/ds000105' 
[INFO   ]  Cloning (counting objects) 100 objects done in 0 sec at 1880853 objects/sec 
[INFO   ]  Cloning (compressing objects) 68466 objects done in 0.001 sec at 52508176 objects/sec 
[INFO   ]  Cloning (receiving objects) 152614 objects done in 6 seconds at 25248 objects/sec 
[INFO   ]  Cloning (resolving stuff) 80063 objects done in 0.003 sec at 23785845 objects/sec 
[INFO   ] access to dataset sibling "datalad" not auto-enabled, enable with:
| 		datalad siblings -d "/tmp/ds000105" enable -s datalad 
install(ok): /tmp/ds000105 (dataset)
[INFO   ]  derivatives/mriqc/anatomical_sub-3.pdf 63.2 kB done in 0.172 sec at 366.9 kB/sec 
[INFO   ]  derivatives/mriqc/anatomical_sub-4.pdf 63.5 kB done in 0.13 sec at 490.0 kB/sec 
[INFO   ]  derivatives/mriqc/functional_group.pdf 182.8 kB done in 0.045 sec at 4.1 MB/sec 
[INFO   ]  derivatives/mriqc/funcMRIQC.csv 29.0 kB done in 0.198 sec at 146.6 kB/sec 
^C[WARNING] Still have 4 active progress bars when stopping 
[INFO   ]  Total partially (1.38% of 1.9 GB) done in 17 seconds at 1.5 MB/sec 
[INFO   ]  sub-1/anat/sub-1_T1w.nii.gz partially (84.71% of 6.0 MB) done in 15 seconds at 319.8 kB/sec 
[INFO   ]  sub-1/func/sub-1_task-objectviewing_run-02_bold.nii.gz partially (22.58% of 25.8 MB) done in 15 seconds at 373.5 kB/sec 
[INFO   ]  sub-1/func/sub-1_task-objectviewing_run-01_bold.nii.gz partially (28.36% of 25.7 MB) done in 15 seconds at 478.3 kB/sec 
[INFO   ]  sub-1/func/sub-1_task-objectviewing_run-03_bold.nii.gz partially (24.89% of 25.8 MB) done in 14 seconds at 428.5 kB/sec 
ERROR: 
Interrupted by user while doing magic: KeyboardInterrupt() [cmd.py:_process_one_line:358]
DATALAD_UI_PROGRESSBAR=log datalad install -g ///openfmri/ds000105  6.27s user 2.08s system 24% cpu 34.728 total

$> sudo rm -rf QA simon dartmouth-phantoms velasco; DATALAD_UI_PROGRESSBAR=log datalad install -g ///labs/tarr/face_place      
[INFO   ] Cloning http://datasets.datalad.org/labs/tarr/face_place [1 other candidates] into '/tmp/face_place' 
[INFO   ]  Cloning (counting objects) 100 objects done in 0 sec at 2853268 objects/sec 
[INFO   ]  Cloning (compressing objects) 28649 objects done in 0.001 sec at 38002092 objects/sec 
[INFO   ]  Cloning (receiving objects) 31593 objects done in 1.418 sec at 22275 objects/sec 
[INFO   ]  Cloning (resolving stuff) 387 objects done in 0.003 sec at 137477 objects/sec 
install(ok): /tmp/face_place (dataset)
[INFO   ] To obtain some keys we need to fetch an archive of size 42.5 MB 
[INFO   ]  MD5E-s42453590--70418661c05ac77945454b58f7a02789.zip 42.5 MB done in 32 seconds at 1.3 MB/sec 
[INFO   ] To obtain some keys we need to fetch an archive of size 28.8 MB 
^C[WARNING] Still have 1 active progress bars when stopping 
[INFO   ]  Total possibly partially done in 46 seconds at 1.3 MB/sec 
[INFO   ]  MD5E-s28831000--2d7e46964c86637e36c26745b8455aea.zip partially (28.79% of 28.8 MB) done in 6 seconds at 1.2 MB/sec 
ERROR: 
Interrupted by user while doing magic: KeyboardInterrupt() [cmd.py:_process_one_line:358]
DATALAD_UI_PROGRESSBAR=log datalad install -g ///labs/tarr/face_place  10.14s user 2.26s system 20% cpu 1:00.97 total

$> sudo rm -rf QA simon dartmouth-phantoms velasco; DATALAD_UI_PROGRESSBAR=log datalad install -g ///labs/tarr/face_place
[INFO   ] To obtain some keys we need to fetch an archive of size 28.8 MB 
[INFO   ]  MD5E-s28831000--2d7e46964c86637e36c26745b8455aea.zip 28.8 MB done in 14 seconds at 2.0 MB/sec 
[INFO   ] To obtain some keys we need to fetch an archive of size 118.2 MB 
[INFO   ]  MD5E-s118229974--970c5c314ec41a2950e98aeefb2faf44.zip 118.2 MB done in a minute at 1.1 MB/sec 
[INFO   ] To obtain some keys we need to fetch an archive of size 16.7 MB 
[INFO   ]  MD5E-s16663731--4735c5882fd6736c1b0829c428a40503.zip 16.7 MB done in 11 seconds at 1.4 MB/sec 
[INFO   ] To obtain some keys we need to fetch an archive of size 16.3 MB 
[INFO   ]  MD5E-s16251761--d049c7c7316b38e847833f574a1ba2ad.zip 16.3 MB done in 13 seconds at 1.2 MB/sec 
[INFO   ]  Total 40.8 MB done in 2 minutes at 233.6 kB/sec 
action summary:
  get (ok: 5290)
  install (notneeded: 1)
DATALAD_UI_PROGRESSBAR=log datalad install -g ///labs/tarr/face_place  69.89s user 21.58s system 50% cpu 3:02.77 total
```
</details>

Ideally should have dedicated test(s) but not sure if I would have juice for that ATM, but let's see if some bugs get revealed by just using it through the tests

### CODE REVIEW

**Comment by yarikoptic** [comment]
@effigies how does this one look to you? ;)

**Comment by effigies** [comment]
Definitely much cleaner. And it works without any obvious bugs on my side.

File: datalad/interface/common_cfg.py [hunk 1]
```diff
@@ -306,4 +306,11 @@
         'default': 256,
         'type': EnsureInt(),
     },
+    'datalad.ui.progressbar': {
+        'ui': ('question', {
+            'title': 'UI progress bars',
+            'text': 'Which backend for progress reporting to use unless specified otherwise?'}),
```
Comment (kyleam) [inline datalad/interface/common_cfg.py]: Perhaps "Default backend for progress reporting"?

File: datalad/ui/progressbars.py [hunk 2]
```diff
@@ -66,7 +85,85 @@ def __init__(self, label='', fill_text=None, total=None, unit='B', out=sys.stdou
...
 
 
-progressbars = {'silent':  SilentProgressBar}
+class LogProgressBar(ProgressBarBase):
+    """A progress bar which logs upon completion of the item
+    """
+    def __init__(self, *args, **kwargs):
+        super(LogProgressBar, self).__init__(*args, **kwargs)
+        # I think we never generate progress bars unless we are the beginning
+        # of reporting something lengthy.  .start is not always invoked
```
Comment (kyleam) [inline datalad/ui/progressbars.py]: s/are/are at/, though I'm missing the connection between .start being invoked and recording the start time here.

File: datalad/ui/progressbars.py [hunk 3]
```diff
@@ -66,7 +85,85 @@ def __init__(self, label='', fill_text=None, total=None, unit='B', out=sys.stdou
...
+                    if perc_done <= 100:
+                        msg += "partially (%.2f%% of %s) "
+                        args += [
+                            perc_done,
+                            self._naturalsize(self.total)
+                        ]
+                    else:
+                        # well well -- we still probably have some issue with
+                        # over-reporting when getting data from datalad-archives
+                        # Instead of providing non-sense % here, j
```
Comment (kyleam) [inline datalad/ui/progressbars.py]: Incomplete comment?

**Comment by yarikoptic** [comment]
> I was surprised this series made no mention of log_progress. LogProgressBar and log.log_progress do related but different things. It'd be nice if the docstring of LogProgressBar pointed to log.log_progress as a way to get progress bars when attached to a tty but incremental log messages otherwise (as opposed to just the final log message here). And log_progress could point to LogProgressBar as well.

doh -- forgot about it!  will do cross reference.  Ideally we should (not in this PR hopefully) may be unify the msg format if applicable?

**Comment by kyleam** [comment]
> Ideally we should (not in this PR hopefully) may be unify the msg format if applicable?

The final message for log_progress() is entirely up to the caller, so I'm not sure what we would unify.

**Comment by yarikoptic** [comment]
ah ok, thanks for the clarification.  I have pushed my changes to address your comments -- thank you for the review!  so now it would look like
```
[INFO   ] Cloning http://datasets.datalad.org/labs/tarr/face_place [1 other candidates] into '/tmp/face_place' 
[INFO   ]  Cloning (counting objects) 100 objects done in 5.79357e-05 sec at 1726051 objects/sec 
[INFO   ]  Cloning (compressing objects) 28649 objects done in 0.00212502 sec at 13481725 objects/sec 
[INFO   ]  Cloning (receiving objects) 31593 objects done in 1.68729 sec at 18724 objects/sec 
[INFO   ]  Cloning (resolving stuff) 387 objects done in 0.00789118 sec at 49042 objects/sec 
install(ok): /tmp/face_place (dataset)
[INFO   ] To obtain some keys we need to fetch an archive of size 42.5 MB 
[INFO   ]  MD5E-s42453590--70418661c05ac77945454b58f7a02789.zip 42.5 MB done in 35 seconds at 1.2 MB/sec 
[INFO   ] To obtain some keys we need to fetch an archive of size 28.8 MB 
...
``` 
still could be considered a bit "too precise" but IMHO not too bad ;) improve to your liking

**Comment by kyleam** [comment]
Thanks for the updates.  I reworded one of the cross-references.  Feel free to drop that commit if it doesn't look ok to you.

> still could be considered a bit "too precise" but IMHO not too bad ;) improve to your liking

I'd still lop off some digits, but that's both subjective and doesn't matter :]

**Pull request merged**